Ockam node memory usage fixes

spacekookie opened the pull request:
This PR attempts to fix the memory leaks present in the current Inlet/Outlet demo.

Testing data is available here (maybe if you're logged in?): https://miro.com/app/board/uXjVO0PpZLg=/

## Rationale

We create "bare contexts" all over the codebase to send messages, wait for specific replies, etc. These don't have a worker relay behind them and thus are outside of the regular shutdown mechanism. When transferring a 132MB file across Ockam the Outlet router allocated up to 14 THOUSAND addresses that were never cleared.

This change:

1. adds some debug logging so that we can have a better understanding of how many addresses are registered on a system.  In the future we may also want to extend this metrics reporting to numbers of clusters, channel load, etc
2. side-steps the normal shutdown mechanism for bare context records, which don't have a relay and can't interact with the "new" shutdown ACK mechanism.

This seems to fix the memory leak present in the demo.

spacekookie committed:
refactor(rust): attempt to cut down memory usage via context drop

spacekookie committed:
refactor(rust): fix address de-allocation issues for bare contexts

SanjoDeundiak commented on implementations/rust/ockam/ockam_node/src/context.rs:
```diff
@@ -46,6 +47,19 @@ pub struct Context {
 
+impl Drop for Context {
+    fn drop(&mut self) {
+        block_future(&self.rt, async {
```
Shouldn't this call be ignored for non-bare Context?

spacekookie committed:
refactor(rust): add an async_drop mechanism for bare context drop

The current implementation of bare Context Drop requires a future
blocking call in a sync function, which causes issues in some of our
tests (and is generally not very elegant).

This commit adds a simple wrapper around receiving a message via a
Oneshot channel (which doesn't require await and can thus be run from
a sync function), and then maps it onto an async message send to the
router to de-allocate a specific resource.

The API for this structure is currently still very specific, but we
can easily change it in the future to allow for different use-cases.

spacekookie commented:
Okay, after some of the tests didn't like that I used a `block_future` call in the Drop handler I implemented an `AsyncDrop` mechanism, which seems to fix that issue, while also fixing the memory problem.

From the table I linked in the Miro board:

- Transferring a 132M video file at ~850 Kb/s (or KB? I don't know what curl uses as a metric)
- Outlet memory usage peaks at 9.5M. Importantly, sending multiple files doesn't seem to increase this number
- Inlet memory useage peaks at 1.7M (again, doesn't seem to increase with repeated transfers)

twittner commented on implementations/rust/ockam/ockam_node/src/context.rs:
```diff
@@ -37,15 +39,29 @@ impl AddressType {
 
+impl Drop for Context {
+    fn drop(&mut self) {
+        if let Some(sender) = core::mem::replace(&mut self.async_drop_sender, None) {
```
```suggestion
        if let Some(sender) = self.async_drop_sender.take() {
```

spacekookie replied:
Oh yea okay. I keep forgetting about `Option::take` :joy:

spacekookie replied:
Oh yea okay. I keep forgetting about `Option::take` :joy:

SanjoDeundiak commented on implementations/rust/ockam/ockam_node/src/context.rs:
```diff
@@ -1,3 +1,5 @@
+use crate::async_drop::AsyncDrop;
+// use crate::block_future;
```
Please remove

SanjoDeundiak commented on implementations/rust/ockam/ockam_node/src/context.rs:
```diff
@@ -136,13 +153,24 @@ impl Context {
 
     async fn new_context_impl(&self, addr: Address) -> Result<Context> {
```
Because lifecycle of bare vs worker Context differs, it would be great to have easily distinguished functions to create each type. AFAIK, for worker, Context::new is used, while Context::new_context_impl is only used for bare variant. But this is very hard to notice without deep investigation + it may change in the future which will lead to bugs.

spacekookie replied:
Hmm, yea. I agree that this is confusing. Although having had a look at the way we construct `Context` I didn't see a super obvious alternative.

We could have `Context::new` and `Context::new_bare` but the functions would be almost identical (except for the async_drop field). Maybe it would be enough to pass in a value for `async_drop_sender` to `Context::new` so that it becomes more obvious how the two types of Context are different?

SanjoDeundiak commented on implementations/rust/ockam/ockam_node/src/context.rs:
```diff
@@ -99,6 +115,7 @@ impl Context {
                 sender,
                 address,
                 mailbox,
+                async_drop_sender: None,
```
Same comment: not obvious that we only create worker's Context here, so that drop handle is None

SanjoDeundiak commented on implementations/rust/ockam/ockam_node/src/router/stop_worker.rs:
```diff
@@ -9,10 +9,12 @@ use ockam_core::{Address, Result};
 pub(super) async fn exec(
     router: &mut Router,
     addr: &Address,
+    bare: bool,
```
Hm... Any other options for the name? Detached? Autonomous?

spacekookie replied:
I created a thread on slack to find a better name. I do agree that "bare context" is very ambiguous :sweat_smile:

SanjoDeundiak reviewed (changes requested):
Great idea!

spacekookie committed:
refactor(rust): rename new_context to new_detached

Since BARE encoding and "bare context" is a naming clash, this commit
changes the latter to "detached context".  We add a type alias, and
changing all references in the code from bare to detached to make the
concept less confusing.

spacekookie commented:
Woops I had forgotten to add the label. Does the bot still pick up the PR? :thinking:
